README.md
# pinyin2ipa

[![dependencies Status](https://david-dm.org/Connum/npm-pinyin2ipa/status.svg)](https://david-dm.org/Connum/npm-pinyin2ipa) [![devDependencies Status](https://david-dm.org/Connum/npm-pinyin2ipa/dev-status.svg)](https://david-dm.org/Connum/npm-pinyin2ipa?type=dev) [![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)  [![npm total downloads](https://img.shields.io/npm/dt/pinyin2ipa.svg)](https://www.npmjs.com/package/pinyin2ipa)

Converts Mandarin Chinese Pinyin notation to IPA (International Phonetic Alphabet) notation. Supports different methods of transliteration, by default using an approach by the Beijing Language and Culture University (北京语言大学).


# Installation
`npm install pinyin2ipa`

or use the files in `/dist` or from the [releases](https://github.com/Connum/npm-pinyin2ipa/releases) for direct in-browser usage.

# Usage and examples
```js
const pinyin2ipa = require('pinyin2ipa').default
// or
import pinyin2ipa from 'pinyin2ipa'
// or (for usage in a browser environment)
<script src="dist/pinyin2ipa.min.js"></script>

// and then it's as simple as
pinyin2ipa("nĭ hăo"); // result: ni3 xau3

// in addition to diacritics, number notation can also be used,
// with spaces or without
pinyin2ipa("ni3hao3"); // result: ni3 xau3
pinyin2ipa("ni3 hao3"); // result: ni3 xau3

// to use an alternative built-in method,
// set the "method" option to "sophisticated"
pinyin2ipa("nĭ hăo", {
    method: "sophisticated"
}); // result: ni3 xɑʊ̯3

// In order to define a custom method, extend the object in pinyin2ipa.methods.
// A method is an object whose keys represent the pinyin notation without
// diacritics, optionally followed by a tone number if that tone needs to be
// treated differently. The value represents the IPA transliteration.
// See the built-in methods in src/methods/ for a better understanding.
pinyin2ipa.methods.myCustomMethod = {
    // ...
    'den': 'tən',
    'de5': 'tə',
    'de': 'tɤ',
    // ...
}
// and then use it like this:
pinyin2ipa('nĭhăoma', {
    method: 'myCustomMethod'
}) 
```

For more examples, see `examples/examples.js` or the tests in `test/index.js`.

## Options
Options can be passed via an object as the second argument of `pinyin2ipa()` (see code example above).

| Option Name  | Default | Description |
| ------------- | ------------- | ------------- |
| `method`  | `"default"`  | Either `"default"` or `"sophisticated"`, or a custom method defined via `pinyin2ipa.methods`.<br><br>`"default"` is an approach used by the Beijing Language and Culture University (北京语言大学)<br>`"sophisticated"` uses an alternative method with more diacritics and some different characters to resemble the actual pronunciation even more closely.<br><br>For defining custom methods, see the code example above. |
| `toneMarker`  | `"number"`  | The tone marker to use after each syllable (neutral tone will not be marked at all by default, see `markNeutral` option). Either `"number"`, `"chaonumber"`, or `"chaoletter"`.<br><br>`"number"` uses integers from 1 to 4<br>`"chaonumber"` uses the Chao tone numerals 55, 35, 214, and 51<br>`"chaoletter"` uses the Chao tone letters ˥, ˧˥, ˨˩˦, and ˥˩
| `markNeutral`  | `false`  | Whether to mark the neutral tone. Only has an effect if `toneMarker` is `number`.
| `superscript`  | `false`  | Boolean indicating whether to display numbers as superscript (¹²³⁴⁵) or normal numbers (12345). Only has an effect if the `toneMarker` option is either `"number"` or `"chaonumber"`. |
| `filterUnknown`  | `true`  | Boolean indicating whether to filter out any unknown syllables. If `false`, unknown syllables will be surrounded by asterisks |

# Development Commands
- `npm run clean` - Remove `lib/` directory
- `npm test` - Run tests with linting
- `npm test:only` - Run tests without linting
- `npm test:watch` - Re-run tests on file changes
- `npm test:prod` - Run tests with minified code
- `npm run test:examples` - Test examples with node
- `npm run lint` - Run ESlint with airbnb-config
- `npm run build` - Babel will transpile ES6 => ES5 and minify the code, Browserify will create a bundle in `dist/` for in-browser usage.
- `npm run prepublish` - Hook for npm. Do all the checks before publishing the module.

# License

MIT © Connum

based on [flexdinesh/npm-module-boilerplate](https://github.com/flexdinesh/npm-module-boilerplate), MIT © Dinesh Pandiyan


CHANGELOG.md
# Changelog

## [1.0.4] - 2022-07-19

### Added
* Changelog

### Changed
* log a test call to pinyin2ipa() to the console in test/test_bundle.html
* updated pinyin-separate and all other dependencies except eslint and eslint-config-airbnb due to major changes

[1.0.4]: https://github.com/Connum/npm-pinyin2ipa/compare/1.0.2...1.0.4


test/index.js
import { assert } from 'chai';
import pinyin2ipa from '../src';

describe('Conversion tests', () => {
  it('should convert nĭhăoma', () => {
    const expectedVal = 'ni3 xau3 ma';
    assert.equal(pinyin2ipa('nĭhăoma'), expectedVal);
  });

  it('should convert nĭhăoma mith markNeutral option', () => {
    const expectedVal = 'ni3 xau3 ma5';
    assert.equal(pinyin2ipa('nĭhăoma', {
      markNeutral: true
    }), expectedVal);
  });

  it('should convert nĭhăoma mith markNeutral option and chao numbers', () => {
    const expectedVal = 'ni214 xau214 ma';
    assert.equal(pinyin2ipa('nĭhăoma', {
      toneMarker: 'chaonumber',
      markNeutral: true
    }), expectedVal);
  });

  it('should convert nĭ hăo ma', () => {
    const expectedVal = 'ni3 xau3 ma';
    assert.equal(pinyin2ipa('nĭ hăo ma'), expectedVal);
  });

  it('should convert nĭhăoma with superscript', () => {
    const expectedVal = 'ni³ xau³ ma';
    assert.equal(pinyin2ipa('nĭhăoma', {
      superscript: true
    }), expectedVal);
  });

  it('should convert nĭhăoma with sophisticated method and superscript chao numbers', () => {
    const expectedVal = 'ni²¹⁴ xɑʊ̯²¹⁴ mɑ';
    assert.equal(pinyin2ipa('nĭhăoma', {
      method: 'sophisticated',
      toneMarker: 'chaonumber',
      superscript: true
    }), expectedVal);
  });

  it('should convert nĭhăoma with sophisticated method and chao letters', () => {
    const expectedVal = 'ni˨˩˦ xɑʊ̯˨˩˦ mɑ';
    assert.equal(pinyin2ipa('nĭhăoma', {
      method: 'sophisticated',
      toneMarker: 'chaoletter'
    }), expectedVal);
  });

  it('should convert ni3hao3ma with superscript numbers', () => {
    const expectedVal = 'ni³ xau³ ma';
    assert.equal(pinyin2ipa('ni3hao3ma', {
      toneMarker: 'number',
      superscript: true
    }), expectedVal);
  });

  it('should convert ni3 hao3 ma with superscript chao numbers', () => {
    const expectedVal = 'ni²¹⁴ xau²¹⁴ ma';
    assert.equal(pinyin2ipa('ni3 hao3 ma', {
      toneMarker: 'chaonumber',
      superscript: true
    }), expectedVal);
  });

  it('should convert dé dè de de5', () => {
    const expectedVal = 'tɤ2 tɤ4 tə tə';
    assert.equal(pinyin2ipa('dé dè de de5'), expectedVal);
  });

  it('should convert dé dè de de5 with markNeutral option', () => {
    const expectedVal = 'tɤ2 tɤ4 tə5 tə5';
    assert.equal(pinyin2ipa('dé dè de de5', {
      markNeutral: true
    }), expectedVal);
  });

  it('should convert ni3 foo bar baz with filterUnknown', () => {
    const expectedVal = 'ni3';
    assert.equal(pinyin2ipa('ni3 foo bar baz', {
      filterUnknown: true
    }), expectedVal);
  });

  it('should convert ni3 foo bar baz without filterUnknown', () => {
    const expectedVal = 'ni3 *foo* *bar* *baz*';
    assert.equal(pinyin2ipa('ni3 foo bar baz', {
      filterUnknown: false
    }), expectedVal);
  });

  it('should convert nĭ\\r\\nhăo\\nma with line breaks', () => {
    const expectedVal = 'ni3\nxau3\nma';
    assert.equal(pinyin2ipa('nĭ\r\nhăo\nma'), expectedVal);
  });
});

describe('Custom method test', () => {
  it('should convert nĭmenhăoma ménhào with custom method', () => {
    const expectedVal = 'foo3 bar baz3 bat lorem2 ipsum4';
    pinyin2ipa.methods.myCustomMethod = {
      ni: 'foo',
      men: 'lorem',
      men5: 'bar',
      hao: 'ipsum',
      hao3: 'baz',
      ma: 'bat'
    };
    assert.equal(pinyin2ipa('nĭmenhăoma ménhào', {
      method: 'myCustomMethod'
    }), expectedVal);
  });
});


src/index.js
import pinyinSeparate from 'pinyin-separate';
import defaultMethod from './methods/default.json';
import sophisticatedMethod from './methods/sophisticated.json';

const defaultOptions = {
  method: 'default',
  toneMarker: 'number',
  markNeutral: false,
  superscript: false,
  filterUnknown: true
};

const toneReps = {
  number: ['1', '2', '3', '4', '5'],
  numberSuperscript: ['¹', '²', '³', '⁴', '⁵'],
  chaonumber: ['55', '35', '214', '51', ''],
  chaonumberSuperscript: ['⁵⁵', '³⁵', '²¹⁴', '⁵¹', ''],
  chaoletter: ['˥', '˧˥', '˨˩˦', '˥˩', '']
};

const pinyinRegEx = /^(((miu|[pm]ou|[bpm](o|e(i|ng?)?|a(ng?|i|o)?|i(e|ng?|a[no])?|u))|(f(ou?|[ae](ng?|i)?|u))|(d(e(i|ng?)|i(a[on]?|u))|[dt](a(i|ng?|o)?|e(i|ng)?|i(a[on]?|e|ng|u)?|o(ng?|u)|u(o|i|an?|n)?))|(neng?|[ln](a(i|ng?|o)?|e(i|ng)?|i(ang|a[on]?|e|ng?|u)?|o(ng?|u)|u(o|i|an?|n)?|ve?))|([ghk](a(i|ng?|o)?|e(i|ng?)?|o(u|ng)|u(a(i|ng?)?|i|n|o)?))|(zh?ei|[cz]h?(e(ng?)?|o(ng?|u)?|ao|u?a(i|ng?)?|u?(o|i|n)?))|(song|shua(i|ng?)?|shei|sh?(a(i|ng?|o)?|en?g?|ou|u(a?n|o|i)?|i))|(r([ae]ng?|i|e|ao|ou|ong|u[oin]|ua?n?))|([jqx](i(a(o|ng?)?|[eu]|ong|ng?)?|u(e|a?n)?))|((a(i|o|ng?)?|ou?|e(i|ng?|r)?))|(w(a(i|ng?)?|o|e(i|ng?)?|u))|y(a(o|ng?)?|e|in?g?|o(u|ng)?|u(e|a?n)?))[0-5]?)+$/i;

function removeDiacritics(word) {
  return word
    .replace(/[āáǎăà]/ig, 'a')
    .replace(/[ēéěěĕè]/ig, 'e')
    .replace(/[īíǐĭì]/ig, 'i')
    .replace(/[ōóǒŏò]/ig, 'o')
    .replace(/[ūúǔŭù]/ig, 'u')
    .replace(/([ǖǘǚǚü̆ǜv̄v́v̆v̌v̀]|[ūúǔŭù]:)/ig, 'ü');
}

function numberifyTones(wordIn) {
  let word = wordIn;
  if (/\d$/.test(word)) {
    return word;
  }

  if (/[āēīōūǖv̄]/.test(word)) {
    word += '1';
  } else if (/[áéíóúǘv́]/.test(word)) {
    word += '2';
  } else if (/[ǎăěĕǐĭǒŏǔŭǚǚü̆v̆v̌]/.test(word)) {
    word += '3';
  } else if (/[àèìòùǜv̀]/.test(word)) {
    word += '4';
  } else {
    word += '5';
  }
  return removeDiacritics(word);
}

function isNumberPinYin(wordIn) {
  let word = wordIn;
  word = word.replace(/\r?\n/g, '');
  return pinyinRegEx.test(word);
}

function isPinYin(wordIn) {
  let word = wordIn;
  word = removeDiacritics(word.replace(/\r?\n/g, ''));
  return pinyinRegEx.test(word);
}

function pinyinWord2IPA(wordIn, methodTable, toneRep, options) {
  let word = wordIn;
  if (word === '#~linebreak~#') return '\n';

  if (parseInt(word, 10) === word) {
    return word;
  }

  if (!isNumberPinYin(word)) {
    const splits = pinyinSeparate(word);
    const convertedWords = [];
    if (splits.length > 1) {
      let part = null;
      let convertedWord = null;
      for (let i = 0; i < splits.length; i++) {
        part = splits[i];
        convertedWord = pinyinWord2IPA(part, methodTable, toneRep, options);
        if (convertedWord.length) {
          convertedWords.push(convertedWord);
        }
      }
      return convertedWords.join(' ');
    }
    word = numberifyTones(word.toLowerCase());
  }

  const pureWord = word.replace(/([^\d]+)\d+/g, '$1');
  if (!isPinYin(pureWord)) {
    return options.filterUnknown ? '' : `*${pureWord}*`;
  }

  const toneMatch = word.match(/\d$/);
  const toneIndex = (toneMatch ? parseInt(toneMatch[0], 10) : 5) - 1;
  const toneMark = toneIndex === 4 && (!options.markNeutral || options.toneMarker !== 'number') ? '' : toneRep[toneIndex];
  if (pureWord === '\\n') return '\n';

  let result = '';
  if (word === pureWord && methodTable[`${word}5`]) {
    result = methodTable[`${word}5`] + toneMark;
  } else if (methodTable[word]) {
    result = methodTable[word] + toneMark;
  } else if (methodTable[pureWord]) {
    result = methodTable[pureWord] + toneMark;
  } else {
    result = options.filterUnknown ? '' : `*${pureWord}*`;
  }

  return result;
}

function pinyin2ipa(pinyIn, optionsArg = defaultOptions) {
  let options = optionsArg;
  let output = pinyIn;

  if (options !== defaultOptions) {
    options = Object.assign({}, defaultOptions, options);
  }

  if (options.superscript && options.toneMarker !== 'number' && options.toneMarker !== 'chaonumber') {
    options.superscript = false;
  }

  const toneRep = toneReps[`${options.toneMarker}${options.superscript ? 'Superscript' : ''}`];

  output = output.replace(/(\d+)([^ ])/g, '$1 $2');
  output = output.replace(/\r?\n/g, ' #~linebreak~# ');

  let words = output.split(/[ '".,;:?!*“”\-—$§€()/´`]+/g);
  words = words.filter(w => !/^ +$/.test(w));

  const methodTable = pinyin2ipa.methods[options.method];
  words = words.map(word => pinyinWord2IPA(word, methodTable, toneRep, options));

  words = words.filter(w => !/^ +$/.test(w));

  return words.join(' ').replace(/ ?(#~linebreak~#|\n) ?/g, '\n').trim();
}

pinyin2ipa.methods = {
  default: defaultMethod,
  sophisticated: sophisticatedMethod
};

export default pinyin2ipa;

// export { pinyin2ipa };
